semant.ml

```diff
@@ -1,156 +1,271 @@
 open Ast
 open Sast
-
 module StringMap = Map.Make (String)
 
-let check (program) =
-  let funcs = program.decls.funcs
-  in
+(* Semantic checking of the AST. Returns an SAST if successful,
+   throws an exception if something is wrong.
+
+   Check each global variable, then check each function *)
+
+let check program =
+  let globals = program.decls.vars in
+  let funcs = program.decls.funcs in
+
   let check_binds (kind : string) (binds : bind list) =
-    List.iter (function
-      (Null, b) -> raise (Failure ("illegal null " ^ kind ^ " " ^ b))
-    | _ -> ()) binds;
-  let rec dups = function
-      [] -> ()
-    |	((_,n1) :: (_,n2) :: _) when n1 = n2 -> raise (Failure ("duplicate " ^ kind ^ " " ^ n1))
-    | _ :: t -> dups t
-    in dups (List.sort (fun (_,a) (_,b) -> compare a b) binds)
+    List.iter
+      (function
+        | Null, b -> raise (Failure ("illegal null " ^ kind ^ " " ^ b))
+        | _ -> ())
+      binds;
+    let rec dups = function
+      | [] -> ()
+      | (_, n1) :: (_, n2) :: _ when n1 = n2 ->
+          raise (Failure ("duplicate " ^ kind ^ " " ^ n1))
+      | _ :: t -> dups t
+    in
+    dups (List.sort (fun (_, a) (_, b) -> compare a b) binds)
+  in
+
+  (**** Check global variables ****)
+  check_binds "global" globals;
+
+  (**** Check functions ****)
+  let built_in_decls =
+    let add_bind map (name, ty) =
+      StringMap.add name
+        {
+          return = Null;
+          fname = name;
+          formals = [ (ty, "PLACEHOLDER") ];
+          stmts = [];
+        }
+        map
+    in
+    List.fold_left add_bind StringMap.empty
+      [ 
+        ("print", Int); 
+        ("printf", Float);
+        ("printb", Bool);
+        ("printm", Matrix);
+        ("printmf", Matrix);
+        ("rows", Matrix);
+        ("cols", Matrix);

```

Comment:
@HuaxuanGAO Could you help me out on this? We need to declare these predefined functions with their argument types here, but I couldn't figure out the argument type for "rows" and "cols". See failing test "test-matrix_dim"

Reply:
Nvm, I figured out. We need to also define the return type for these functions.